samples/gibbs_ensemble/gibbs.py

```diff
@@ -0,0 +1,139 @@
+            pos=np.random.random((n, 3)) * self._system.box_l)
+
+    def remove_last_added_particle(self):
+        '''Removes the last added particle (reverts insert particle).'''
+        for p in self.last_added_particles: p.remove()
+
+    def energy(self):
+        return self._system.analysis.energy()["total"]
+
+    def _recv_data(self):
+        '''Receives data and return it.'''
+        # The first 4 bytes encode the length of the messages received.
+        buf = b''
+        while len(buf) < 4:
+            buf += self._socket.recv(4 - len(buf))
+        length = struct.unpack('!I', buf)[0]
+        msg = pickle.loads(self._socket.recv(length))
+        return msg
+
+    def _send_data(self, data):
+        '''Send the data packet.'''
+        # The first 4 bytes encode the length of the messages sent.
+        length = struct.pack('>I', len(data))
+        packet = length + data
+        self._socket.send(packet)
+
+    def run(self, host, port):
+        # init socket
+        self._socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self._socket.connect((host, port))
+        msg = self._recv_data()
+        if msg[0] != MessageId.START:
+            raise Exception("Invalid message received from controller")
+
+        # send the initial energy
+        energy = self.energy()
+        self._send_data(pickle.dumps([MessageId.ENERGY, energy]))
+
+        while msg[0] != MessageId.END:
+            # receive command to execute next step
+            msg = self._recv_data()
+            if msg[0] == MessageId.END:
+                break
+            elif msg[0] == MessageId.MOVE_PART:
+                self.move_particle()
+                self._send_data(pickle.dumps(
+                    [MessageId.ENERGY, self.energy()]))
+            elif msg[0] == MessageId.CHANGE_VOLUME:
+                box_l = msg[1]
+                self.change_volume_and_rescale_particles(box_l)
+                self._send_data(pickle.dumps(
+                    [MessageId.ENERGY, self.energy()]))
+            elif msg[0] == MessageId.CHANGE_VOLUME_REVERT:
+                box_l = msg[1]
+                self.change_volume_and_rescale_particles(box_l)
+            elif msg[0] == MessageId.EXCHANGE_PART_ADD:
+                n = msg[1]
+                self.insert_particle(n)
+                self._send_data(pickle.dumps(
+                    [MessageId.ENERGY, self.energy()]))
+            elif msg[0] == MessageId.EXCHANGE_PART_REMOVE:
+                self.remove_particle()
+                self._send_data(pickle.dumps(
+                    [MessageId.ENERGY, self.energy()]))
+            elif msg[0] == MessageId.MOVE_PART_REVERT:
+                self.move_particle_revert()
+            elif msg[0] == MessageId.EXCHANGE_PART_ADD_REVERT:
+                self.remove_last_added_particle()
+            elif msg[0] == MessageId.EXCHANGE_PART_REMOVE_REVERT:
+                self.remove_particle_revert()
```

Comment:
this whole block is probably something like a ```MessageHandler``` that can react on messages of type ```MessageId``` and is more or less just a dict from ```MessageId``` to a function.

Reply:
I implemented this. Tutors will have to do a bit more explaining when teching Python beginners, though.